src/modules/bookings/bookings.service.ts

```diff
@@ -0,0 +1,190 @@
+import {
+  BadRequestException,
+  Inject,
+  Injectable,
+  NotFoundException,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { CreateBookingReqDto, UpdateBookingStatusReqDto } from './dto/request';
+import { Booking } from './entities/booking.entity';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { BookingsQueryParamsReqDto } from './dto/request/bookings-query-params.dto';
+import { BookingStatus } from './constants/booking-status.constant';
+import { BOOKING_SERVICE_FEE } from './constants/booking-service-fee.constant';
+import { BookingStatusTransitions } from './constants/bookings-status-transition.constant';
+import { AccommodationsService } from '../accommodations/accommodations.service';
+
+@Injectable()
+export class BookingsService {
+  constructor(
+    @InjectRepository(Booking)
+    private readonly bookingRepository: Repository<Booking>,
+    @Inject(AccommodationsService)
+    private readonly accommodationsService: AccommodationsService,
+  ) {}
+
+  async getMyBookings(
+    queryParamsDto: BookingsQueryParamsReqDto,
+    userId: string,
+  ) {
+    const queryBuilder = this.bookingRepository.createQueryBuilder('booking');
+
+    queryBuilder.where('booking.userId = :userId', { userId });
+
+    if (queryParamsDto.status) {
+      queryBuilder.andWhere('booking.status = :status', {
+        status: queryParamsDto.status,
+      });
+    }
+
+    queryBuilder
+      .leftJoinAndSelect('booking.accommodation', 'accommodation')
+      .select(['booking', 'accommodation']);
+
+    return await queryBuilder.getMany();
+  }
+
+  async getBooking(userId: string, bookingId: string): Promise<Booking> {
+    const booking = await this.bookingRepository.findOne({
+      where: { id: bookingId },
+      relations: ['accommodation', 'user', 'accommodation.owner'],
+    });
+
+    if (
+      booking.accommodation.owner.id !== userId &&
+      booking.user.id !== userId
+    ) {
+      throw new UnauthorizedException(
+        'You are not allowed to view this booking',
+      );
+    }
+
+    if (!booking) {
+      throw new NotFoundException(`Booking with id ${bookingId} not found`);
+    }
+
+    return booking;
+  }
+
+  async createBooking(
+    dto: CreateBookingReqDto,
+    userId: string,
+  ): Promise<Booking> {
+    const accommodation = await this.accommodationsService.getAccommodationById(
+      dto.accommodationId,
+    );
+
+    if (!accommodation) {
+      throw new NotFoundException(
+        `Accommodation with id ${dto.accommodationId} not found`,
+      );
+    }
+
+    if (accommodation.owner.id === userId) {
+      throw new BadRequestException("You can't book your own accommodation");
+    }
+
+    if (dto.checkOutDate <= dto.checkInDate) {
```

Comment:
Since you receive dates as strings, you should do something like this `new Date(dto.checkInDate).valueOf() <= new Date(dto.checkOutDate).valueOf()` to avoid miscomparison. Same with other date comparisons as well. Better create a util that does this.

Reply:
they are converted to dates during the dto validation